carcara/src/checker/parallel/scheduler/mod.rs

```diff
@@ -0,0 +1,210 @@
+    used_by: HashSet<usize>,
+}
+
+impl<'a> StackLevel<'a> {
+    pub fn new(id: usize, cmds: &'a [ProofCommand], pre_req: Option<(usize, usize)>) -> Self {
+        Self {
+            id,
+            cmds,
+            pre_req,
+            used_by: HashSet::new(),
+        }
+    }
+}
+
+/// Struct that stores the schedules for each thread.
+pub struct Scheduler {
+    pub loads: Vec<Schedule>,
+}
+
+impl Default for Schedule {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Scheduler {
+    /// Creates a thread scheduler for this proof using a specific number of
+    /// workers. This scheduler is responsible for balancing the load (the
+    /// proof steps have different costs to be checked) aiming the minimum
+    /// amount of async overhead
+    pub fn new(num_workers: usize, proof: &Proof) -> (Self, Vec<usize>) {
+        // Initializes the control and result variables
+        let cmds = &proof.commands;
+        let mut loads = vec![Schedule::new(); num_workers];
+        let mut stack = vec![StackLevel::new(0, cmds, None)];
+        let mut pq = BinaryHeap::<AssignedLoad>::new();
+        let mut context_usage = vec![];
+        for i in 0..num_workers {
+            pq.push(AssignedLoad(0, i));
+        }
+
+        loop {
+            // Pop the finished subproofs
+            while !stack.is_empty() && {
+                let top = stack.last().unwrap();
+                top.id == top.cmds.len()
+            } {
+                for schedule_id in &stack.last().unwrap().used_by {
+                    let last = loads[*schedule_id].last().unwrap();
+                    // If it's an useless context insertion
+                    if last.0 < stack.len()
+                        && matches!(stack[last.0].cmds[last.1], ProofCommand::Subproof(_))
+                    {
+                        // Make sure this context usage count is reduced
+                        let subproof_id = match &stack[last.0].cmds[last.1] {
+                            ProofCommand::Subproof(s) => s.context_id,
+                            _ => unreachable!(),
+                        };
+                        context_usage[subproof_id] -= 1;
+
+                        loads[*schedule_id].pop();
+                    }
+                    // Creates a closing step for each schedule that used this subproof
+                    else {
+                        loads[*schedule_id].push((stack.len() - 1, usize::MAX));
+                    }
+                }
+                stack.pop();
+            }
+            if stack.is_empty() {
+                break;
+            }
+            //
+            let AssignedLoad { 0: mut load, 1: load_index } = pq.pop().unwrap();
+            {
+                let top = stack.last().unwrap();
+                let step_weight = get_step_weight(&top.cmds[top.id]);
+                assert!(u64::MAX - step_weight >= load, "Weight balancing overflow!");
+                load += step_weight;
```

Comment:
You can use `checked_add` to protect against overflows:
```suggestion
                load = load
                    .checked_add(step_weight)
                    .expect("Weight balancing overflow!");
```